core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -967,7 +967,16 @@ class Partition(val topicPartition: TopicPartition,
     }
   }
 
-  def appendRecordsToLeader(records: MemoryRecords, origin: AppendOrigin, requiredAcks: Int): LogAppendInfo = {
+  /**
+   * @param completeDelayedRequests It may requires a bunch of group locks when completing delayed requests so it may
```

Comment:
the caller of ```ReplicaManager#appendRecords``` may hold the group lock so it could produce deadlock if ```ReplicaManager#appendRecords``` tries to complete  purgatory.

Reply:
read @hachikuji https://github.com/apache/kafka/pull/8657#discussion_r427432140 again. It is a nice idea to refactor ```ReplicaManager``` and ```Partition``` to simplify the behavior of checking delayed operations.

Could I address the refactor in another PR to avoid bigger patch?

Reply:
Yes, we can refactor that in a separate PR. Could you file a followup jira for that?

Reply:
https://issues.apache.org/jira/browse/KAFKA-10170